Add example of creating tools using HoloViz Panel and HoloViews

MarcSkovMadsen opened the pull request:
## Description

Adds an example of creating tools using HoloViz Panel and Holoviews requested in #6023

https://user-images.githubusercontent.com/42288570/141610857-f42eb134-c521-46e7-b2a2-23162aca382e.mp4

## Todo

- [x] Add Panel and HoloViews to dependencies
- [x] Fix pep8 issues
- [ ] Find a way to build and test docs when on Windows
- [ ] Test that example works

## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

Commit:
add first version of holoviz example

pep8speaks commented:
Hello @MarcSkovMadsen! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`doc/examples/applications/plot_tool_using_holoviz.py`](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py):

> [Line 7:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L7): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 10:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L10): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (140 > 79 characters)
> [Line 12:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L12): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)
> [Line 15:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L15): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 19:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L19): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (117 > 79 characters)
> [Line 86:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L86): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 88:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L88): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)
> [Line 96:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L96): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 110:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L110): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 119:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L119): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 156:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L156): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 175:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L175): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (256 > 79 characters)
> [Line 177:80](https://github.com/scikit-image/scikit-image/blob/6f7594620baa8eb4af126c9079c13b3b6a9c2f6d/doc/examples/applications/plot_tool_using_holoviz.py#L177): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (166 > 79 characters)

Commit:
fix pep8 issues

Commit:
add dependencies

MarcSkovMadsen commented:
I'm not able to install latex on my constrained, windows laptop and thus cannot build the docs and test them.

I've tried using binder without being able to get a development environment up and running.

If somebody knows a working, easy way to get a dev environment up and running please let me know. Thanks.

mkcor commented:
Dear @MarcSkovMadsen,

Thank you for this great contribution. I have fetched your branch locally and was able to build the docs successfully! :tada: 

Your addition renders like this:
![new_tutorial](https://user-images.githubusercontent.com/2227806/141981893-1a143912-e107-45cd-9f54-25617bbab300.png)

In fact, with [Sphinx-Gallery](https://sphinx-gallery.github.io/stable/index.html), the syntax to use in your Python files is [reStructuredText](https://sphinx-gallery.github.io/stable/syntax.html#embedding-rst) (not Markdown). For example, for external links:
https://github.com/scikit-image/scikit-image/blob/fa43626bd6c3db027ae339b7ecbef9f433070010/doc/examples/applications/plot_3d_image_processing.py#L228

Do you mind updating your contribution accordingly? Then I'm happy to review it thoroughly!

MarcSkovMadsen commented:
Thanks @mkcor. Will do 👍

MarcSkovMadsen commented:
I cannot include an `.mp4` in the document. But I can include a `.gif`. For now I upload here to get a `url` to include.

![scikit-image-w-holoviz](https://user-images.githubusercontent.com/42288570/146666639-02f0106f-a3e2-4306-9033-707554e4e61e.gif)

Commit:
convert markdown to restructuredText

Commit:
added links to more inspiration

MarcSkovMadsen commented:
Thanks @mkcor 

Thanks for your patience. I got distracted. It's updated now. Have a nice Christmas.

Commit:
fix misspelling

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,189 @@
+.. figure:: https://user-images.githubusercontent.com/42288570/146666639-02f0106f-a3e2-4306-9033-707554e4e61e.gif
+   :alt: A video showcasing the application
+
+   A video showcasing the application
```
```suggestion
   Screencast showcasing the example application.

```

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,189 @@
+"""
+=======================================
+Interactive Tools and Apps with HoloViz
```
```suggestion
Explore and annotate image data interactively with with HoloViz
```
Please write titles at the imperative and without capitalizing non-first words. This convention also applies to section titles. Thanks!

PS: How do 'HoloViz' and 'HoloViews' relate?

MarcSkovMadsen replied:
HoloViz provides a set of packages. HoloViews is one such package. Panel is another such package. See https://holoviz.org/

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,189 @@
+Interactive Tools and Apps with HoloViz
+=======================================
+
+This tutorial shows how to build interactive tools and applications
```
```suggestion
This tutorial shows how to build interactive dashboards and applications

```

MarcSkovMadsen replied:
I've changed. But I would actually like to change it back as "tool" is important. To me a tool is some smaller interactive component that you can use in a notebook or in a web application. For your self or your colleagues. And Panel is really good at that. That is one of the reasons is my preferred framework over the competitors. A dashboard is a very visual application with some interactivity. And application can be a dashboard. But it could also be a chat application or something 3rd. Panel is actually really great for smaller chat applications :-).

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,189 @@
+=======================================
+
+This tutorial shows how to build interactive tools and applications
+using `Panel <https://panel.holoviz.org>`_, `HoloViews <https://holoviews.org>`_ you can use to make you and your teams
+analysis more interactive and powerful.
```
```suggestion
using `Panel <https://panel.holoviz.org>`_ and `HoloViews <https://holoviews.org>`_.
These tools can make your analyses more interactive and, hence, powerful.
```

mkcor replied:
After this very general introduction, please introduce what the example application is about, so the reader knows where we're headed!

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,189 @@
+analysis more interactive and powerful.
+
+.. figure:: https://user-images.githubusercontent.com/42288570/146666639-02f0106f-a3e2-4306-9033-707554e4e61e.gif
+   :alt: A video showcasing the application
```
```suggestion
   :alt: Screencast showcasing the example application.
```

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,189 @@
+from skimage import data, filters
+
+#############################################
+# Activate the Panel and HoloViews extensions
```
```suggestion
# Activate the Panel and HoloViews extensions
# ==================================
```
(looks like this should be a section title, hierarchically at the same level as "Import packages")

mkcor replied:
I think that all three above 'sections' (set up env, import packages, and activate extensions) could/should go under one single section.

mkcor reviewed (commented):
Thank you, @MarcSkovMadsen!

Please improve the writeup along the lines of this quick review. The text can be very concise but it should flow well. Feel free to draw inspiration from well-written gallery examples, such as "Estimate anisotropy in a 3D microscopy image."

MarcSkovMadsen committed:
Update doc/examples/applications/plot_tool_using_holoviz.py

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

MarcSkovMadsen committed:
Update doc/examples/applications/plot_tool_using_holoviz.py

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

MarcSkovMadsen committed:
Update doc/examples/applications/plot_tool_using_holoviz.py

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

MarcSkovMadsen committed:
Update doc/examples/applications/plot_tool_using_holoviz.py

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

MarcSkovMadsen committed:
Update doc/examples/applications/plot_tool_using_holoviz.py

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

MarcSkovMadsen committed:
Update doc/examples/applications/plot_tool_using_holoviz.py

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

Commit:
get inspiration from plot_3d_structure_tensor

Commit:
add suggestions from review

MarcSkovMadsen commented:
Thanks @mkcor . I've tried to apply the inspiration from `plot_3d_structure_tensor.py`. I hope you can provide me some more guidance. Thanks.

mkcor commented:
PS: @MarcSkovMadsen did you know you could apply suggested changes as a *batch*? https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request

MarcSkovMadsen commented:
Thx. I did not know.

mkcor commented on requirements/docs.txt:
```diff
@@ -20,3 +20,5 @@ myst-parser
 plotly>=4.14.0
 kaleido
+panel==0.12.4
+holoviews==1.14.6
```
Can you please relax the dependency versions so that they are just lower bounds, instead of exact matches?

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,181 @@
+"""
+===============================================================
+Explore and annotate image data interactively with with HoloViz
```
```suggestion
Explore and annotate image data interactively with HoloViz
```

mkcor reviewed (commented):
Please proofread your text and please wrap your lines at 79 characters. Let me know when you're done and I'll review! Thanks.

MarcSkovMadsen committed:
fix spelling error

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

Commit:
relax requirements

Commit:
respect 79 character limit

MarcSkovMadsen commented:
I've relaxed the versions, fixed the spelling error, tried to identify more spelling errors, updated to respect 79 character limit.

In two places I've not been able to respect the 79 character limit as 

I've tried to google to find out how I can break lines in `figure` and `code` block without luck. So I've kept them as they are.

![image](https://user-images.githubusercontent.com/42288570/154807421-e023397f-6a56-44eb-99f9-2bd5346f58f3.png)

Please let me know if there is more to do. Thanks.

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,202 @@
+Explore and annotate image data interactively with HoloViz
+==========================================================
+
+In this tutorial, we build interactive tools and an application using
+`Panel <https://panel.holoviz.org>`_ and `HoloViews <https://holoviews.org>`_.
```
```suggestion
In this tutorial, we build interactive tools using
`Panel <https://panel.holoviz.org>`_ and `HoloViews <https://holoviews.org>`_,
two visualization libraries maintained by `HoloViz <https://holoviz.org/>`_. 
```
Since 'HoloViz' appears in the title, it must appear in the body text at some point, preferably towards the beginning, so the reader knows what we are talking about...

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,202 @@
+
+In this tutorial, we build interactive tools and an application using
+`Panel <https://panel.holoviz.org>`_ and `HoloViews <https://holoviews.org>`_.
+
+We will build an application based on the Sobel algorithm to interactively
+find a great color map for the visualization. We will add some nice features,
+like a download button, to make sharing your findings easier.
```
```suggestion
We build a web application based on the Sobel algorithm which lets the user
find interactively a great color map for the visualization.
Then we add interaction features such as a download button.
```
* Keep the same tense.
* What is "finding a great color map [for the visualization]" supposed to mean? Even if we're not dealing with a scientific problem, the question should be somewhat *interesting*, so the reader can see the point of building this app... From your animation, it looks like the app is only about *changing* the color map interactively (not really *finding*... some kind of optimum, as the verb 'finding' implies). If the app is 'based on the Sobel algorithm' as you write, then what seems 'interesting' would be for the user to vary the `mode` and/or `cval` values...
* I think everyone knows what a download button is for... To make the comment relevant, it should specify what exactly is being downloaded upon clicking the download button in question: a CSV of the full dataset? a CSV of the subset currently displayed? a PNG of the static picture? I think it would be more fitting to specify this when we are at that point in the tutorial.

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,202 @@
+   Screencast showcasing the example application.
+
+
+You can develop using notebooks or .py files. You can use it with a very
```
What does 'it' refer to?

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,202 @@
+We will build an application based on the Sobel algorithm to interactively
+find a great color map for the visualization. We will add some nice features,
+like a download button, to make sharing your findings easier.
+
+These techniques can make your analyses more interactive and, hence, powerful.
```
```suggestion
Interactive applications provide flexible ways to visualize and explore data,
making data analysis more powerful and, at the same time, accessible to
non-technical audiences.
```
You might as well stick with a very generic sentence.

mkcor commented:
> In the two places below, I've not been able to respect the 79 character limit. I've tried to google to find out how I can break lines in figure and code block without luck. So I've kept them as they are.

Sure, that's fine.

mkcor commented on doc/examples/applications/plot_tool_using_holoviz.py:
```diff
@@ -0,0 +1,202 @@
+    width=width,
+    cmap="binary_r",
+    title="Before",
+    active_tools=["box_zoom"],
```
@MarcSkovMadsen could we do without option 'active_tools' so that we could use a backend other than 'bokeh' and, hence/perhaps, have the figure display in the static HTML page?

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
